ctxgroup: panic in Wait() if task panics

dt opened the pull request:
Previously if a caller were to pass some piece of code to a ctxgroup to execute and then wait on that execution, that caller would not see any panics in the execution of the task, as it is on a different goroutine.

For example, consider some function that does two steps and handles panics:

```
func f(ctx) {
  defer func() { if r := recover(); r != nil { handle(r) })()
  foo()
  bar()
}
```

A refactor may then opt to run foo and bar concurrently:

```
func f(ctx) {
  defer func() { if r := recover(); r != nil { handle(r) })()
  g := ctxgroup.WithContext(ctx)
  g.GoCtx(foo)
  g.GoCtx(bar)
  g.Wait()
}
```

However, when `foo` and `bar` were handed to the group, their panics are no longer visible to f and its recover/handling. Additionally, stack traces from these panics are also harder to attribute to f.

This change extends the ctxgroup to recover panics in tasks passed to it just long enough to transport them back to the group, where they are then re-thrown when Wait() is called. In our example above, f() would once again see the panic thrown in foo or bar be thrown when it calls Wait().

In general, this should make the behavior of panics, and what is or is not recovered or handles more predictable when code introduces uses of a ctxgroup, as the same panics should generally continue to be thrown on the same goroutines, instead of elsewhere.

Release note: none.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110393)
<!-- Reviewable:end -->

stevendanna commented on pkg/util/ctxgroup/ctxgroup.go:
```diff
@@ -162,9 +169,33 @@ func (g Group) Go(f func() error) {
+}
+
 // GoCtx calls the given function in a new goroutine.
+// If the function panics before another function has returned an error, calling
+// Wait() will panic with the original panic wrapped in ErrTaskPanic so as to
+// capture its original stack. NB: if another task in the group has already
+// returned an error, the panic will instead be recovered and ignored as the
+// error slot in the underlying errgroup is already used.
```
I wonder if, in dev builds we still want to raise this panic somehow?

dt replied:
alternatively: https://github.com/cockroachdb/cockroach/commit/1a26d107acf9b9758a6db0fade53783b95d9c431

dt replied:
I messed with this a little and now just store the panic myself, along-side the errgroup instead of inlining the errgroup.

ecwall commented:
This PR looks good to me. Is merging it a prereq for https://github.com/cockroachdb/cockroach/issues/110346? Is there anything else that would block that from being done?

dt commented:
> Is merging it a prereq for https://github.com/cockroachdb/cockroach/issues/110346

I believe so. I don't want to go adding recovers to the root goroutines in jobs/schedules when so much of the work is performed on other goroutines if panics on those other goroutines won't actually get back to that root recover; doing that would just mean we have inconsistent and unpredictable story about when a panic does or doesn't crash the node. So I think I want to start by ensuring any panic anywhere in a job execution crashes -- and only crashes -- the root goroutine job/schedule goroutine, and then we can add recovers to those two.

ecwall commented on pkg/util/ctxgroup/ctxgroup_test.go:
```diff
@@ -42,3 +43,17 @@ func TestErrorAfterCancel(t *testing.T) {
+		panic(1)
+	}
+
+	require.Panics(t, func() {
```
Can you assert that the outer panic contains the inner panic's stacktrace?

dt replied:
Done.

knz reviewed (dismissed):
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110393)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @ecwall, @miretskiy, and @stevendanna)

___
*[`pkg/util/ctxgroup/ctxgroup.go` line 173 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-Nfkj28h66KU724eFlSv:-Nfkj28h66KU724eFlSw:bxdmmin) ([raw file](https://github.com/cockroachdb/cockroach/blob/64bba367a6d76dc260357b9918a32deb29f03414/pkg/util/ctxgroup/ctxgroup.go#L173)):*
> ```Go
> 		wrapped: grp,
> 		ctx:     ctx,
> 		panicMu: &recovered{},
> ```

Any reason why `recovered` is included by pointer?
___
*[`pkg/util/ctxgroup/ctxgroup.go` line 251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-Nfkir-1A3nq1I1lkkSh:-Nfkir-1A3nq1I1lkkSi:b-nx6rv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/64bba367a6d76dc260357b9918a32deb29f03414/pkg/util/ctxgroup/ctxgroup.go#L251)):*
> ```Go
> 		e = errors.WithStackDepth(err, depth+1)
> 	} else {
> 		e = errors.NewWithDepthf(depth+1, "panic: %v", r)
> ```

This is not good -- if the go runtime panics with a non-error object, we generally don't want to turn it into an error. These panics are not recoverable.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110393)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ecwall, @knz, @miretskiy, and @stevendanna)

___
*[`pkg/util/ctxgroup/ctxgroup.go` line 173 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-Nfkj28h66KU724eFlSv:-Nfly2lb8X5qOLMgYFba:b-mos3dv) ([raw file](https://github.com/cockroachdb/cockroach/blob/64bba367a6d76dc260357b9918a32deb29f03414/pkg/util/ctxgroup/ctxgroup.go#L173)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Any reason why `recovered` is included by pointer?
</blockquote></details>

`Group` is commonly passed by value (due to embedding); it's other fields are *errgroup and ctx, which are both pointer types, which makes this safe, but means I need a pointer type around my mu/mu protected field here to maintain these semantics.
___
*[`pkg/util/ctxgroup/ctxgroup.go` line 251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-Nfkir-1A3nq1I1lkkSh:-NflynNwBPNUpor6Qivw:b-pctt6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/64bba367a6d76dc260357b9918a32deb29f03414/pkg/util/ctxgroup/ctxgroup.go#L251)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This is not good -- if the go runtime panics with a non-error object, we generally don't want to turn it into an error. These panics are not recoverable.
</blockquote></details>

my goal here isn't so much to decide what I can and cannot recover here, but rather to _transport_ the panic from the fn in the group to the Wait(). 

A previous iteration of this I actually didn't wrap it in error at all -- I just used `interface{}` to pass it from Go to Wait to rethrow the exact same value -- same pointer -- that was thrown in `Go`. Unfortunately in discussions with Evan/jobs people, we don't like that if we do so, we lose the original stack trace. We want the stack trace from where we originally panicked to be preserved and visible when we re-throw the panic in Wait(). Wrapping it in an error if it isn't already seemed like the best way to get this.

As far as panics that aren't errors, there are plenty that aren't errors in our codebase as well, e.g. `panic(fmt.Sprintf("external storage provider already registered for %s", providerType.String()))` or `panic(fmt.Sprintf(`batches cannot have capacity larger than %d; requested %d`, max, capacity))` or `panic(fmt.Errorf("JobDetailsForEveryJobType has the incorrect entry for type %s", typ))`, and we've seen similar ones come out of 3rd party libraries that we use in jobs, so I'm confident that to implement our goal here, of moving the panic from `Go` to `Wait` we need to move panics where the payload isn't `error`.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
Seems reasonable to me. Let's resolve this discussion about non-error panics, but other than that this seems useful.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-:-NfoqAIk81k2nDHGt9u_:b14zofj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @ecwall, @knz, and @miretskiy)

___
*[`pkg/util/ctxgroup/ctxgroup.go` line 251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-Nfkir-1A3nq1I1lkkSh:-NfooRK5B3bi_yHfwv9B:bbci3tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/64bba367a6d76dc260357b9918a32deb29f03414/pkg/util/ctxgroup/ctxgroup.go#L251)):*
> These panics are not recoverable.

Can we say more about which panics we are worried about? I believe if we hit a _truly_ non-recoverable runtime panic from the runtime itself, the runtime converts it to a "throw" that isn't recovered and directly exits.

https://github.com/golang/go/blob/352c8835e7609ad72872b5a63bffeb6142b29d98/src/runtime/panic.go#L1011-L1019

https://github.com/golang/go/blob/352c8835e7609ad72872b5a63bffeb6142b29d98/src/runtime/panic.go#L1201

https://github.com/golang/go/blob/master/src/runtime/panic.go#L55-L80

But, this is the first time I've peaked at this code so perhaps I'm unfamiliar with the category of panics you are concerned with.
___
*[`pkg/util/ctxgroup/ctxgroup.go` line 192 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-NfooFwC4pn-0IAVR52b:-NfooFwC4pn-0IAVR52c:b-8vd8xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a9dfe21fd776ff732605031d0d066317c45ca78/pkg/util/ctxgroup/ctxgroup.go#L192)):*
> ```Go
> 				defer g.panicMu.Unlock()
> 				g.panicMu.payload = errors.CombineErrors(g.panicMu.payload, err)
> 			}
> ```

[nit] I could see putting the interior of this if block in a method hanging off *recover.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110393)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @ecwall, @miretskiy, and @stevendanna)

___
*[`pkg/util/ctxgroup/ctxgroup.go` line 251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-Nfkir-1A3nq1I1lkkSh:-Nfor4UPEFDu-vxCmWIs:b-ml61lt) ([raw file](https://github.com/cockroachdb/cockroach/blob/64bba367a6d76dc260357b9918a32deb29f03414/pkg/util/ctxgroup/ctxgroup.go#L251)):*
> Unfortunately in discussions with Evan/jobs people, we don't like that if we do so, we lose the original stack trace. We want the stack trace from where we originally panicked to be preserved and visible when we re-throw the panic in Wait().

I am friendly to this ask, but the solution to this is **not** to convert the object into an error. Take a stack trace (into a string), combine it with the original object (possibly with a custom struct and a suitable `Format` function), then throw that.

> Can we say more about which panics we are worried about?

Seen in the wild: accessing slices out of bounds; mutex lock errors due to internal corruption of the G or P structs; CGo calls failing due to invalid FFI struct data, and others. Generally, we do not want to let the process continue to run past these.

> As far as panics that aren't errors, there are plenty that aren't errors in our codebase as well

Most of these are outright bugs / tech debt and should be fixed, many by converting this to `panic(errors.AssertionFailedf())`. This deserves a linter.

> we've seen similar ones come out of 3rd party libraries that we use in jobs,

If we use a 3rd party dependency that throws a panic upon cases that can be encountered in produiction, we need to stop using that dependency and/or fix it. That's also outright a bug.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110393)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ecwall, @knz, and @miretskiy)

___
*[`pkg/util/ctxgroup/ctxgroup.go` line 251 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/110393#-Nfkir-1A3nq1I1lkkSh:-NfpWGg-Bi2oPKBzRpk4:bcejec8) ([raw file](https://github.com/cockroachdb/cockroach/blob/64bba367a6d76dc260357b9918a32deb29f03414/pkg/util/ctxgroup/ctxgroup.go#L251)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

> Unfortunately in discussions with Evan/jobs people, we don't like that if we do so, we lose the original stack trace. We want the stack trace from where we originally panicked to be preserved and visible when we re-throw the panic in Wait().

I am friendly to this ask, but the solution to this is **not** to convert the object into an error. Take a stack trace (into a string), combine it with the original object (possibly with a custom struct and a suitable `Format` function), then throw that.

> Can we say more about which panics we are worried about?

Seen in the wild: accessing slices out of bounds; mutex lock errors due to internal corruption of the G or P structs; CGo calls failing due to invalid FFI struct data, and others. Generally, we do not want to let the process continue to run past these.

> As far as panics that aren't errors, there are plenty that aren't errors in our codebase as well

Most of these are outright bugs / tech debt and should be fixed, many by converting this to `panic(errors.AssertionFailedf())`. This deserves a linter.

> we've seen similar ones come out of 3rd party libraries that we use in jobs,

If we use a 3rd party dependency that throws a panic upon cases that can be encountered in produiction, we need to stop using that dependency and/or fix it. That's also outright a bug.
</blockquote></details>

We discussed this elsewhere and concluded that there is no convention around the type of the panic payload, specifically it implementing `error` or not, and whether that indicates whether we can recover and move on from it or not. We looked through a number of examples from the std library that suggest that "misusage" panics, that indicate a bug in the calling code, can be both errors or plain strings. That said, the runtime itself has a convention, as Steven mentioned above, of using `throw` rather than panic to terminate the program when it encounters what it considers "unrecoverable" error conditions (corruption of stacks, P/M/G structs, etc).

There _is_ a concern about wrapping arbitrary panic objects and changing the object type (for example to make them errors)  _in general_, as doing so in, for example, a func passed in to a library could cause that library's internal panic-throwing/panic-handling to be broken by an unexpected change in type between those two calls. That concern, however, appears to not apply in this specific ctxgroup case: we're calling a function on a brand new goroutine. so we know there is no recovery "above" us that we'd be breaking, so any library-internal panic handling has already had its chance to run by the time this recover is hit.


<!-- Sent from Reviewable.io -->

stevendanna commented:
@dt Perhaps now that the branch is cut we can let this bake on master?

dt committed:
ctxgroup: panic in Wait() if task panics

Previously if a caller were to pass some piece of code to a ctxgroup to
execute and then wait on that execution, that caller would not see any
panics in the execution of the task, as it is on a different goroutine.

For example, consider some function that does two steps and handles panics:

```
func f(ctx) {
  defer func() { if r := recover(); r != nil { handle(r) })()
  foo()
  bar()
}
```

A refactor may then opt to run foo and bar concurrently:

```
func f(ctx) {
  defer func() { if r := recover(); r != nil { handle(r) })()
  g := ctxgroup.WithContext(ctx)
  g.GoCtx(foo)
  g.GoCtx(bar)
  g.Wait()
}
```

However, when A and B were handed to the group, their panics are no longer
visible to f and its recover/handling. Additionally, stack traces from
these panics are also harder to attribute to f.

This change extends the ctxgroup to recover panics in tasks passed to it
just long enough to transport them back to the group, where they are then
re-thrown when Wait() is called. In our example above, f() would once again
see the panic thrown in A or B be thrown when it calls Wait().

In general, this should make the behavior of panics, and what is or is
not recovered or handles more predictable when code introduces uses of a
ctxgroup, as the same panics should generally continue to be thrown on
the same goroutines, instead of elsewhere.

Release note: none.
Epic: none.

dt commented:
bors r+
